File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.java
class Solution {
    //helper function to calculate Days required
    private int daysRequired(int [] weights, int capacity){
        int days=1, currentLoad=0;
        //iterte over all package weights
        for(int w: weights){
            //check if adding current package exceeds capacity
            if(currentLoad +w>capacity){
                // If yes, increase days count since we start a new day
                days++;
                //reset current load to current package weight
                currentLoad=w;

            }else{
                 // Else, add current package weight to current load
                currentLoad += w;
            }
        }
        return days;


    }
    public int shipWithinDays(int[] weights, int days) {
        //first we need to find range
        int low = Arrays.stream(weights).max().getAsInt();
        int high = Arrays.stream(weights).sum();
        //Binary Search
        while(low<=high){
            int mid = low + (high-low)/2;
            //Calculate how many days needed for capacity
            int needed = daysRequired(weights, mid);
            if(needed<=days){
                high = mid-1;
            }else{
                low = mid+1;
            }

        }return low;
    }
}

File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.py
class Solution(object):
    def daysRequired(self, weights, capacity):
        days=1
        currentLoad=0
        for w in weights:
            if currentLoad+w>capacity:
                days+=1
                currentLoad=w
            else:
                currentLoad+=w
        return days

    def shipWithinDays(self, weights, days):
        """
        :type weights: List[int]
        :type days: int
        :rtype: int
        """
        low = max(weights)
        high = sum(weights)
        
        while(low<=high):
            mid = low +(high -low)//2
            needed = self.daysRequired(weights, mid)
            if needed<= days:
                high = mid-1
            else:
                low= mid+1
            
        return low
